mac/Focalboard/ViewController.swift

```diff
@@ -64,20 +78,70 @@ class ViewController:
 	@objc func onServerStarted() {
 		NSLog("onServerStarted")
 		DispatchQueue.main.asyncAfter(deadline: .now() + 0.5) {
-			self.updateSessionToken()
+			self.updateSessionTokenAndUserSettings()
 			self.loadHomepage()
 		}
 	}
 
-	private func updateSessionToken() {
+	private func persistUserSettings() {
+		let semaphore = DispatchSemaphore(value: 0)
+
+		// Convert to base64 to avoid escaping issues when later inserting the exported settings into the import script
+		webView.evaluateJavaScript("window.btoa(Focalboard.exportUserSettings())") { result, error in
+			defer { semaphore.signal() }
+			guard let base64 = result as? String, let data = Data(base64Encoded: base64), let decoded = String(data: data, encoding: .utf8) else {
+				NSLog("Failed to export user settings: \(error?.localizedDescription ?? "?")")
+				return
+			}
+			UserDefaults.standard.set(base64, forKey: "localStorage")
+			NSLog("Persisted user settings: \(decoded)")
+		}
+
+		// During shutdown the system grants us about 5 seconds to clean up and store user data
+		let timeout = DispatchTime.now() + .seconds(3)
+		var result: DispatchTimeoutResult?
+
+		// Busy wait because evaluateJavaScript can only be called from *and* signals on the main thread
+		while (result != .success && .now() < timeout) {
+			result = semaphore.wait(timeout: .now())
+			RunLoop.current.run(mode: .default, before: Date())
+		}
+
+		if result == .timedOut {
+			NSLog("Timed out trying to persist user settings")
+		}
+	}
+
+	private func updateSessionTokenAndUserSettings() {
 		let appDelegate = NSApplication.shared.delegate as! AppDelegate
-		let script = WKUserScript(
+		let sessionTokenScript = WKUserScript(
 			source: "localStorage.setItem('focalboardSessionId', '\(appDelegate.sessionToken)');",
 			injectionTime: .atDocumentStart,
 			forMainFrameOnly: true
 		)
+		let base64 = UserDefaults.standard.string(forKey: "localStorage") ?? ""
+		let userSettingsScript = WKUserScript(
+			source: """
+				const settings = window.atob("\(base64)");
+				if (typeof(Focalboard) !== 'undefined') {
+					if (Focalboard.importUserSettings(settings)) {
+						window.webkit.messageHandlers.\(messageHandlerName).postMessage({
+							type: '\(MessageType.didImportUserSettings.rawValue)',
+							settings: settings
+						});
+					} else {
+						window.webkit.messageHandlers.\(messageHandlerName).postMessage({
+							type: '\(MessageType.didNotImportUserSettings.rawValue)'
+						});
+					}
+				}
+				""",
+			injectionTime: .atDocumentEnd,
```

Comment:
Thanks @Johennes for this PR! I smoke tested it and it appears to work seamlessly (didn't notice a refresh). Some quick thoughts on how we might avoid this reload:
* What if the Swift code injected the base-64 settings string in a placeholder property (e.g. NativeApp.settingsBlob), atDocumentStart.
* Then add code to the initialization routine (somewhere in app.tsx maybe, as long as it's before rendering)
* It would check to see if Native.settingsBlob is set, if so, it would call `importUserSettings`, then delete the blob

Have to think a bit about security (I don't think that's an issue), but the extra benefit is the handling of the blob would be all in shared TS code.

What do you think?

Reply:
This is a great idea! I've been able to drastically reduce the amount of JS code in the native app this way and there's no more page reload necessary. 🎉 

Security-wise, I'm not an expert but I believe this at least doesn't make things worse. The injected blob is accessible by any other script running on the page but it's essentially the same data that's already accessible globally via the `exportUserSettingsBlob` function.